#!/usr/bin env python3
# Snakefile for reprocessing the merged genotype vcf for all donors located in
#/lustre/scratch123/hgi/projects/hipsci/releases/data/gtarray/releases/merged_files/REL-2018-01
# per chromosome
# snakemake is in /software/teamtrynka/conda/trynka-base/bin/snakemake
import os
import numpy as np

POOL=["pool_15"]
CHR = [i for i in np.arange(1,23,1)] + ["X"]

rule all:
    input:
        genotype=expand("../../data/{pool}/{pool}.merged.genotype.exonfiltered.nonIdentical.hg38.vcf.gz",pool=POOL)

rule pool_donor_names:
    input:
        metadata="../../data/sampleMetadata.txt",
        donor_ids="../../data/donor_id_match.txt"
    output:
        donor_pools="../../data/{pool}/{pool}_donorNames.txt"
    message: "Extracting donors present in the pools. Run with snakemake --jobs 100 --cluster [comma] bsub {params.group} {params.queue} {params.threads} {params.memory} {params.jobname} {params.error} [comma]"
    params:
        group= "-G teamtrynka",
        queue="-q normal",
        threads="-n 2",
        memory="-M10000 -R'span[hosts=1] select[mem>10000] rusage[mem=10000]'",
        jobname= "-o ../../logs/log_pool_donor_ids.%J.%I",
        error="-e ../../errors/error_pool_donor_ids.%J.%I"
    shell:
        """
        echo "Subsetting file names to merge, from all hipsci files available looking at the metadata IDs"
        detectedDonors="../../data/{wildcards.pool}/{wildcards.pool}_donorNames.txt"
        cat {input.metadata} | grep -w {wildcards.pool} | awk '{{print $3}}' | tr ';' '\n' > $detectedDonors
        """

rule subset_donors_vcf:
    input:
        donor_pools="../../data/{pool}/{pool}_donorNames.txt",
        all_hipsci_vcfs="../../data/all_pools_no_curn_iukl/hipsci_ipmar_imputed_UK10K_1KG_HRC.MAF01.highqual.hg38.vcf.gz"
    output:
        genotype="../../data/{pool}/{pool}.genotype.vcf.gz",
        index="../../data/{pool}/{pool}.genotype.vcf.gz.csi"
    message: "Subsetting donors present in the metadata file."
    params:
        group= "-G teamtrynka",
        queue="-q normal",
        threads="-n 32",
        memory="-M100000 -R'span[hosts=1] select[mem>100000] rusage[mem=100000]'",
        jobname= "-o ../../logs/log_{pool}_subset_donors_vcf.%J.%I",
        error="-e ../../errors/error_{pool}_subset_donors_vcf.%J.%I"
    shell:
        """
        # Subset donors
        /software/teamtrynka/conda/trynka-base/bin/bcftools view -S {input.donor_pools} {input.all_hipsci_vcfs} --threads 32 -Oz -o {output.genotype}
        /software/teamtrynka/conda/trynka-base/bin/bcftools index {output.genotype}

        """

rule exonfilter_genotype_file:
    input:
        genotype="../../data/{pool}/{pool}.genotype.vcf.gz"
    output:
        filtered="../../data/{pool}/{pool}.genotype.exonfiltered.vcf.gz"
    message: "Subsetting genotypes to exonic positions."
    params:
        group= "-G teamtrynka",
        queue="-q normal",
        threads="-n 20",
        memory="-M40000 -R'span[hosts=1] select[mem>40000] rusage[mem=40000]'",
        jobname= "-o ../../logs/log_exonfilter_genotype_file.%J.%I",
        error="-e ../../errors/error_exonfilter_genotype_file.%J.%I",
        bind_dir="/lustre/scratch123/hgi/mdt1/projects/otar2065/hipsci_genotype_processing/data/{pool}"
    shell:
        """
        echo "Remove chr"
        # remove chr to first column
        # create conversion file beforehand
        # https://www.biostars.org/p/98582/#176041
        /software/teamtrynka/conda/trynka-base/bin/bcftools annotate --rename-chrs ../../data/{wildcards.pool}/rm_chr.txt {input.genotype} > ../../data/{wildcards.pool}/temp_nochr.vcf
        ~/bin/bcftools-1.18/bcftools sort --max-mem 40G --temp-dir ../../data/ --write-index -Oz -o ../../data/{wildcards.pool}/temp_nochr_sorted_presubset.vcf.gz ../../data/{wildcards.pool}/temp_nochr.vcf

        echo "Subset to exon"
        singularity exec --bind {params.bind_dir}/ \
        /software/teamtrynka/Demuxafy.sif vcftools \
        --gzvcf {params.bind_dir}/temp_nochr_sorted_presubset.vcf.gz \
        --max-alleles 2 \
        --remove-indels \
        --bed {params.bind_dir}/genotypes/hg38exonsUCSC.bed \
        --recode \
        --recode-INFO-all \
        --out {params.bind_dir}/{wildcards.pool}.genotype.exonfiltered.nochr

        # add chr to first column
        # create conversion file beforehand
        # https://www.biostars.org/p/98582/#176041
        /software/teamtrynka/conda/trynka-base/bin/bcftools annotate --rename-chrs ../../data/{wildcards.pool}/add_chr.txt \
        -Oz -o ../../data/{wildcards.pool}/{wildcards.pool}.genotype.exonfiltered.vcf.gz \
        ../../data/{wildcards.pool}/{wildcards.pool}.genotype.exonfiltered.nochr.recode.vcf
        /software/teamtrynka/conda/trynka-base/bin/bcftools index -f {params.bind_dir}/{wildcards.pool}.genotype.exonfiltered.vcf.gz
        """

rule subset_variants:
    input:
        genotype="../../data/{pool}/{pool}.genotype.exonfiltered.vcf.gz"
    output:
        genotype="../../data/{pool}/{pool}.merged.genotype.exonfiltered.nonIdentical.hg38.vcf.gz"
    message: "Subsetting to variants that have non-identical genotypes across donors.Shoudln't remove anything after MAF filters."
    params:
        group= "-G teamtrynka",
        queue="-q normal",
        threads="-n 1",
        memory="-M160000 -R'span[hosts=1] select[mem>160000] rusage[mem=160000]'",
        jobname= "-o ../../logs/log_{pool}_subset_variants.%J.%I",
        error="-e ../../errors/error_{pool}_subset_variants.%J.%I"
    shell:
        """
        echo "Subsetting to variants that have non-identical genotypes across donors"
        /software/teamtrynka/conda/otar2065/bin/Rscript 0.Retain_non-identical_variants.R {input.genotype} \
        ../../data/{wildcards.pool}/{wildcards.pool}.merged.genotype.exonfiltered.nonIdentical.hg38.vcf.gz

        # Decompressing and re-compressing in right format
        gunzip -d ../../data/{wildcards.pool}/{wildcards.pool}.merged.genotype.exonfiltered.nonIdentical.hg38.vcf.gz
        /software/teamtrynka/conda/trynka-base/bin/bgzip -f ../../data/{wildcards.pool}/{wildcards.pool}.merged.genotype.exonfiltered.nonIdentical.hg38.vcf

        echo "Generating index file"
        /software/teamtrynka/conda/trynka-base/bin/bcftools index {output.genotype}
        """
