#!/usr/bin/env python3
# Snakefile for proliferation GWAS
# snakemake is in /software/hgi/envs/conda/teamtrynka/ma23/snakemake
# code based on Marta's code in otar2065/OTAR2065_phenotypic_QTLs/code/otar2065_phenotypic_qtls/phagocytosis_associations_for_Daianna/Snakefile

import os
import numpy as np
import snakemake.io
from snakemake.io import *

## Test with a few chunks first
# CHUNK = ['1', '2', '3', '4', '5', '6', '7', '8']
# CHUNK = glob_wildcards("/lustre/scratch123/hgi/projects/otar2065/OTAR2065_phenotypic_QTLs/data/genotypes/full_genotype/genotype_minor_allele_dosage_{chunk}.csv")
CHUNK = range(1, 582)
# CHUNK = 582

## Define all final outputs to be generated

# rule all:
#   input:
#     line_prop_changes_premac_vs_iPSC_genotype_out = expand("../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/premac_vs_iPSC_data/line_prop_changes_premac_vs_iPSC_genotype_{chunk}.Rdata", chunk = CHUNK),
#     line_prop_changes_old_vs_young_premac_genotype_out = expand("../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/old_vs_young_premac_data/line_prop_changes_old_vs_young_premac_genotype_{chunk}.Rdata", chunk = CHUNK),
#     line_prop_changes_microglia_vs_premac_IFN_genotype_out = expand("../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/microglia_vs_premac_IFN_data/line_prop_changes_microglia_vs_premac_IFN_genotype_{chunk}.Rdata", chunk = CHUNK),
#     line_prop_changes_microglia_vs_premac_LPS_genotype_out = expand("../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/microglia_vs_premac_LPS_data/line_prop_changes_microglia_vs_premac_LPS_genotype_{chunk}.Rdata", chunk = CHUNK),
#     line_prop_changes_microglia_vs_premac_untreated_genotype_out = expand("../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/microglia_vs_premac_untreated_data/line_prop_changes_microglia_vs_premac_untreated_genotype_{chunk}.Rdata", chunk = CHUNK)


rule all:
  input:
    GWAS_premac_vs_iPSC = expand("../../output_data/08_GWAS_proliferation/01_Association_testing/premac_vs_iPSC_results/GWAS_premac_vs_iPSC_results_{chunk}.Rdata", chunk = CHUNK),
    GWAS_old_vs_young = expand("../../output_data/08_GWAS_proliferation/01_Association_testing/old_vs_young_premac_results/GWAS_old_vs_young_premac_results_{chunk}.Rdata", chunk = CHUNK),
    GWAS_microglia_vs_premac_IFN = expand("../../output_data/08_GWAS_proliferation/01_Association_testing/microglia_vs_premac_IFN_results/GWAS_microglia_vs_premac_IFN_results_{chunk}.Rdata", chunk = CHUNK),
  	GWAS_microglia_vs_premac_LPS = expand("../../output_data/08_GWAS_proliferation/01_Association_testing/microglia_vs_premac_LPS_results/GWAS_microglia_vs_premac_LPS_results_{chunk}.Rdata", chunk = CHUNK),
  	GWAS_microglia_vs_premac_untreated = expand("../../output_data/08_GWAS_proliferation/01_Association_testing/microglia_vs_premac_untreated_results/GWAS_microglia_vs_premac_untreated_results_{chunk}.Rdata", chunk = CHUNK)

    
# rule input_data_preparation:
#   input:
#     genotype_minor_allele_dosage_input_path = "/lustre/scratch125/humgen/projects_v2/otar2065/OTAR2065_phenotypic_QTLs/data/genotypes/full_genotype/genotype_minor_allele_dosage_{chunk}.csv"
#   output:
#     line_prop_changes_premac_vs_iPSC_genotype_out_path = "../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/premac_vs_iPSC_data/line_prop_changes_premac_vs_iPSC_genotype_{chunk}.Rdata",
#     line_prop_changes_old_vs_young_premac_genotype_out_path = "../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/old_vs_young_premac_data/line_prop_changes_old_vs_young_premac_genotype_{chunk}.Rdata",
#     line_prop_changes_microglia_vs_premac_IFN_genotype_out_path = "../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/microglia_vs_premac_IFN_data/line_prop_changes_microglia_vs_premac_IFN_genotype_{chunk}.Rdata",
#     line_prop_changes_microglia_vs_premac_LPS_genotype_out_path = "../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/microglia_vs_premac_LPS_data/line_prop_changes_microglia_vs_premac_LPS_genotype_{chunk}.Rdata",
#     line_prop_changes_microglia_vs_premac_untreated_genotype_out_path = "../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/microglia_vs_premac_untreated_data/line_prop_changes_microglia_vs_premac_untreated_genotype_{chunk}.Rdata"
# 
# 	message: "Prepare input data for proliferation GWAS for each chunk of 9999 variants."
# 	params:
# 		group = "-G teamtrynka",
# 		queue = "-q normal",
# 		threads = "-n 1",
# 		memory = "-M4000 -R'span[hosts=1] select[mem>4000] rusage[mem=4000]'",
# 		jobname = "-o logs/00_Data_exploration_processing/log_input_data_preparation.{chunk}.%J.%I",
# 		error = "-e errors/00_Data_exploration_processing/error_input_data_preparation.{chunk}.%J.%I"
# 	shell:
# 	 	"""
# 		 set +o pipefail;
# 
# 			## Prepare input data for GWAS for each variant chunk
# 			Rscript 00_Data_exploration_processing.R \
# 			{input.genotype_minor_allele_dosage_input_path} \
# 			{output.line_prop_changes_premac_vs_iPSC_genotype_out_path} \
# 			{output.line_prop_changes_old_vs_young_premac_genotype_out_path} \
# 			{output.line_prop_changes_microglia_vs_premac_IFN_genotype_out_path} \
# 			{output.line_prop_changes_microglia_vs_premac_LPS_genotype_out_path} \
# 			{output.line_prop_changes_microglia_vs_premac_untreated_genotype_out_path}
# 
# 		"""

rule association_testing:
  input:
    line_prop_changes_premac_vs_iPSC_genotype_input_path = "../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/premac_vs_iPSC_data/line_prop_changes_premac_vs_iPSC_genotype_{chunk}.Rdata",
    line_prop_changes_old_vs_young_premac_genotype_input_path = "../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/old_vs_young_premac_data/line_prop_changes_old_vs_young_premac_genotype_{chunk}.Rdata",
    line_prop_changes_microglia_vs_premac_IFN_genotype_input_path = "../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/microglia_vs_premac_IFN_data/line_prop_changes_microglia_vs_premac_IFN_genotype_{chunk}.Rdata",
    line_prop_changes_microglia_vs_premac_LPS_genotype_input_path = "../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/microglia_vs_premac_LPS_data/line_prop_changes_microglia_vs_premac_LPS_genotype_{chunk}.Rdata",
    line_prop_changes_microglia_vs_premac_untreated_genotype_input_path = "../../output_data/08_GWAS_proliferation/00_Data_exploration_processing/microglia_vs_premac_untreated_data/line_prop_changes_microglia_vs_premac_untreated_genotype_{chunk}.Rdata"

  output:
    GWAS_premac_vs_iPSC_out_path = "../../output_data/08_GWAS_proliferation/01_Association_testing/premac_vs_iPSC_results/GWAS_premac_vs_iPSC_results_{chunk}.Rdata",
    GWAS_old_vs_young_premac_out_path = "../../output_data/08_GWAS_proliferation/01_Association_testing/old_vs_young_premac_results/GWAS_old_vs_young_premac_results_{chunk}.Rdata",
    GWAS_microglia_vs_premac_IFN_out_path = "../../output_data/08_GWAS_proliferation/01_Association_testing/microglia_vs_premac_IFN_results/GWAS_microglia_vs_premac_IFN_results_{chunk}.Rdata",
  	GWAS_microglia_vs_premac_LPS_out_path = "../../output_data/08_GWAS_proliferation/01_Association_testing/microglia_vs_premac_LPS_results/GWAS_microglia_vs_premac_LPS_results_{chunk}.Rdata",
  	GWAS_microglia_vs_premac_untreated_out_path = "../../output_data/08_GWAS_proliferation/01_Association_testing/microglia_vs_premac_untreated_results/GWAS_microglia_vs_premac_untreated_results_{chunk}.Rdata"

  message: "Test for association between the variants in each chunk with the 5 proliferation phenotypes."
  params:
    group = "-G teamtrynka",
  	queue = "-q normal",
  	threads = "-n 1",
  	memory = "-M4000 -R'span[hosts=1] select[mem>4000] rusage[mem=4000]'",
  	jobname = "-o logs/01_Association_testing/log_association_testing.{chunk}.%J.%I",
  	error = "-e errors/01_Association_testing/error_association_testing.{chunk}.%J.%I"
  shell:
    """
  		 set +o pipefail;

  			## Test association for each variant in chunk
  			Rscript 01_Association_testing.R \
  			{input.line_prop_changes_premac_vs_iPSC_genotype_input_path} \
  			{input.line_prop_changes_old_vs_young_premac_genotype_input_path} \
  			{input.line_prop_changes_microglia_vs_premac_IFN_genotype_input_path} \
  			{input.line_prop_changes_microglia_vs_premac_LPS_genotype_input_path} \
  			{input.line_prop_changes_microglia_vs_premac_untreated_genotype_input_path} \
  			{output.GWAS_premac_vs_iPSC_out_path} \
  			{output.GWAS_old_vs_young_premac_out_path} \
  			{output.GWAS_microglia_vs_premac_IFN_out_path} \
  			{output.GWAS_microglia_vs_premac_LPS_out_path} \
  			{output.GWAS_microglia_vs_premac_untreated_out_path}

    """
